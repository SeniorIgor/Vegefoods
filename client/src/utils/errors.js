export class ValidationError extends Error {
	constructor(message, data = null) {
		super(message);

		this.name = "ValidationError";
		this.data = data;
	}
}

export class AuthorizationError extends Error {
	constructor(message, isOldToken) {
		super(message);

		this.name = "AuthorizationError";
		this.isOldToken = isOldToken;
	}
}