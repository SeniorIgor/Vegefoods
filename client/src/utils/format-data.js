import { ONE_HOUR } from '../constants';

const getDeliveryDates = (earlyDate) => {
	const day = 24 * 3600 * 1000;

	return [
		{
			id: 1,
			value: earlyDate,
			name: `${getWeekDay(earlyDate)}, ${formatDate(earlyDate)}`
		},
		{
			id: 2,
			value: earlyDate + day,
			name: `${getWeekDay(earlyDate + day)}, ${formatDate(earlyDate + day)}`
		},
		{
			id: 3,
			value: earlyDate + day * 2,
			name: `${getWeekDay(earlyDate + day * 2)}, ${formatDate(earlyDate + day * 2)}`
		}
	]
};

const getDeliveryTime = () => {
  return [
    {
      id: 1,
      value: 8 * ONE_HOUR,
      name: 'from 8:00 to 13:00',
    },
    {
      id: 2,
      value: 13 * ONE_HOUR,
      name: 'from 13:00 to 18:00',
    },
    {
      id: 3,
      value: 18 * ONE_HOUR,
      name: 'from 18:00 to 23:00',
    }
  ];
};

const formatPrice = (number) => {
	return new Intl.NumberFormat('ru-RU', {
		style: "currency",
		currency: "RUB",
		useGrouping: true,
		minimumFractionDigits: 2,
		maximumFractionDigits: 2
	}).format(number);
}

const discountPrice = (price, discount) => {
	return formatPrice(price * (1 - discount / 100));
}

const formatNumber = (number) => {
	return new Intl.NumberFormat('ru-RU', {
		useGrouping: true,
		minimumFractionDigits: 1,
		maximumFractionDigits: 1
	}).format(number);
}

const getWeekDay = (miliseconds) => {
	const date = new Date(miliseconds);

	let days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];

	return days[date.getDay()];
}

const getEarlyDate = () => {
	const now = new Date();
	return (new Date(now.getFullYear(), now.getMonth(), now.getDate() + 2)).getTime();
}

const formatDate = (miliseconds) => {
	const d = new Date(miliseconds);

	const date = [
		'0' + d.getDate(),
		'0' + (d.getMonth() + 1),
		'' + d.getFullYear()
	].map(component => component.slice(-2)); // взять последние 2 цифры из каждой компоненты

	// соединить компоненты в дату
	return date.slice(0, 3).join('.');
}

const formatDateTime = (miliseconds) => {
	const d = new Date(miliseconds);

	const date = [
		'0' + d.getDate(),
		'0' + (d.getMonth() + 1),
		'' + d.getFullYear(),
		'0' + d.getHours(),
		'0' + d.getMinutes()
	].map(component => component.slice(-2)); // взять последние 2 цифры из каждой компоненты

	// соединить компоненты в дату
	return date.slice(0, 3).join('.') + ' ' + date.slice(3).join(':');
}

const formatPhoneNumber = (value) => {
	const result = '+';

	switch (true) {
		case (value.length === 1):
			return `${result}${value} `;
		case (value.length > 0 && value.length < 5):
			return `${result}${value.slice(0, 1)} (${value.slice(1)}`;
		case (value.length > 4 && value.length < 8):
			return `${result}${value.slice(0, 1)} (${value.slice(1, 4)}) ${value.slice(4)}`;
		case (value.length > 7 && value.length < 10):
			return `${result}${value.slice(0, 1)} (${value.slice(1, 4)}) ${value.slice(4, 7)}-${value.slice(7)}`;
		case (value.length > 9):
			return `${result}${value.slice(0, 1)} (${value.slice(1, 4)}) ${value.slice(4, 7)}-${value.slice(7, 9)}-${value.slice(9)}`;
		default:
			break;
	}

	return result;
}

export {
	formatPrice,
	discountPrice,
	formatNumber,
	getWeekDay,
	getDeliveryDates,
	getDeliveryTime,
	formatDate,
	formatDateTime,
	formatPhoneNumber,
	getEarlyDate,
}