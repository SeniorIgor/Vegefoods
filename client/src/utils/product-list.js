export const getProductsForPage = (pageId, productsPerPage, products) => {
	const pageIndex = (pageId - 1) * productsPerPage;
	return products.slice(pageIndex, pageIndex + productsPerPage);
}
