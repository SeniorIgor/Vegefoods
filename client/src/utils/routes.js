import React from 'react';
import { Switch, Route } from 'react-router-dom';
import ErrorIndicator from '../components/error-indicator';
import {
	HomePage, ProductsPage, ProductPage, ContactsPage, CartPage, OrdersPage,
	FavoritesPage, OrderingPage, OrderingSuccessPage, NotRegisterPage
} from '../components/pages';

export const useRoutes = (isAuthentificated) => {
	if (isAuthentificated) {
		return (
			<Switch>
				<Route path="/" component={HomePage} exact />
				<Route path="/catalog/:category/page/:pageId" component={ProductsPage} exact />
				<Route path="/product/:productId" component={ProductPage} exact />
				<Route path="/cart" component={CartPage} exact />
				<Route path="/ordering" component={OrderingPage} exact />
				<Route path="/ordering-success" component={OrderingSuccessPage} exact />
				<Route path="/orders" component={OrdersPage} exact />
				<Route path="/favorites" component={FavoritesPage} exact />
				<Route path="/contacts" component={ContactsPage} exact />

				{/* <Route path="/about" component={TestPage} exact /> */}
				<Route component={ErrorIndicator} />
			</Switch>
		)
	}

	return (
		<Switch>
			<Route path="/" component={HomePage} exact />
			<Route path="/catalog/:category/page/:pageId" component={ProductsPage} exact />
			<Route path="/product/:productId" component={ProductPage} exact />
			<Route path="/cart" component={NotRegisterPage} exact />
			<Route path="/favorites" component={NotRegisterPage} exact />
			<Route path="/contacts" component={ContactsPage} exact />
			<Route path="/orders" component={NotRegisterPage} exact />
			<Route component={ErrorIndicator} />
		</Switch>
	);
}