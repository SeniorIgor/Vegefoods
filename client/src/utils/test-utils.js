import React from 'react';

export const setUp = (Component, creator, initialProps = {}) =>
	(props = {}) => creator(<Component {...initialProps} {...props} />);