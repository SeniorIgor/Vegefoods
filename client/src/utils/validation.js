
const validEmailRegex = RegExp(/^(([^<>()\[\]\.,;:\s@\"]+(\.[^<>()\[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/i);

const validateForm = (errors) => {
	let valid = true;
	Object.values(errors).forEach(
		(val) => val.length > 0 && (valid = false)
	);
	return valid;
}

const fieldLengthError = (value, fieldName, length) => {
	if (value && value.length >= length) return '';

	return `${fieldName[0].toUpperCase() + fieldName.slice(1)
		} должен содержать не менее ${length} ${length === 1 ? 'символа' : 'символов'
		}`;
}

const emailError = (value) => validEmailRegex.test(value) ? '' : 'Неверный Email!';

const confirmError = (value, password) => password !== value ? 'Пароли должны совпадать!' : '';

export {
	validEmailRegex,
	validateForm,
	fieldLengthError,
	emailError,
	confirmError,
}