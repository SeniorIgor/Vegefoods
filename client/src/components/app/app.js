import React from 'react';

import Newsletter from './../newsletter';
import Footer from './../footer';

import Header from '../header';
import { MainMenuContainer } from '../../containers';
import { withAuth } from '../../components/hoc';
import { useRoutes } from '../../utils';

import './app.scss';

export const App = ({ isAuth }) => {

	const routes = useRoutes(isAuth);

	return (
		<div className="app">
			<Header />
			<MainMenuContainer />
			{routes}
			<Newsletter />
			<Footer />
		</div>
	);
}

export default withAuth(App);