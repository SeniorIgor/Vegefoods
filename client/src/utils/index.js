import { useRoutes } from './routes';
import { ValidationError, AuthorizationError } from './errors';
import {
	validEmailRegex,
	validateForm,
	fieldLengthError,
	emailError,
	confirmError,
} from './validation';
import {
	formatPrice,
	discountPrice,
	formatNumber,
	getWeekDay,
	getEarlyDate,
	getDeliveryDates,
	getDeliveryTime,
	formatDate,
	formatDateTime,
	formatPhoneNumber,
} from './format-data';
import { getProductsForPage } from './product-list';
import { setUp } from './test-utils';


export {
	// format-data
	formatNumber,
	discountPrice,
	formatPrice,
	formatPhoneNumber,
	formatDate,
	formatDateTime,
	// routes
	useRoutes,
	// validation
	validEmailRegex,
	validateForm,
	fieldLengthError,
	emailError,
	confirmError,
	// date
	getWeekDay,
	getEarlyDate,
	getDeliveryDates,
	getDeliveryTime,
	// errors
	ValidationError,
	AuthorizationError,
	// product-list
	getProductsForPage,
	// test-utils
	setUp,
}