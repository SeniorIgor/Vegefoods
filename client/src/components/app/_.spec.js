import React from 'react';
import { shallow } from 'enzyme';

import { App } from './app';
import Header from '../header';
import { setUp } from '../../utils';

const render = setUp(App, shallow, { isAuth: false });

describe('APP: ', () => {

	let wrapper;
	beforeEach(() => {
		wrapper = render();
	})

	it('component contain Header component', () => {
		expect(wrapper.find(Header)).toHaveLength(1);
	})

	it('snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	})

})