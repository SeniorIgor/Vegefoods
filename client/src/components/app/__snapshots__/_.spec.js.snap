// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APP:  snapshot 1`] = `
<div
  className="app"
>
  <Component />
  <Component />
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/catalog/:category/page/:pageId"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/product/:productId"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/cart"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/favorites"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/contacts"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/orders"
    />
    <Route
      component={[Function]}
    />
  </Switch>
  <Newsletter />
  <Footer />
</div>
`;
